WIP: driver: uart: stm32: implement async api

jli157 opened the pull request:
By using DMA transfer, implement UART Async API for STM32F4 boards. So far, it is still under development and some issues still exist:
- [x] Test TX
- [x] Implement missing APIs.
- [ ] validate with `/tests/drivers/uart/uart_async_api`
- [ ] Remove self test application
- [ ] Rebase and clean up

So, the current one is just for review and to get suggestions. 

The test vehicle is a NUCLEO F429ZI board. 

The PR is trying to resolve the issue: #13955, and base on PR #25459 and #25492
Also, parts of the implementation referred to [the implementation](https://github.com/StefJar/zephyr_stm32_uart3_dma_driver) from @StefJar.

jli157 commented:
@StefJar Please review. Thanks!

erwango commented:
@jli157, please use https://github.com/zephyrproject-rtos/zephyr/pull/13364 to complete your work

jli157 commented:
> @jli157, please use #13364 to complete your work

Sure, will work on that.

erwango commented:
@jli157: Can you clarify difference between this PR and #15375?

jli157 commented:
> @jli157: Can you clarify difference between this PR and #15375?

This one is still based on the existing STM32 DMA driver and could be dropped off and the PR #15375 will be the replacement for this one if the PR #13364 is finally merged. Basically, both are same.

erwango commented:
@jli157 , any update on this?

erwango commented:
^^ FRASTM

jli157 commented:
@erwango Sorry I'm on vacation recently. I'll rebase and retest this one after I come back in middle of July.

erwango commented:
@jli157, txs for update, enjoy your vacations!

ioannisg commented:
@jli157 @erwango moving this to 2.1 - doesn't seem this can make it to 2.0 (if yes, pls, switch back to 2.0 milestone)

erwango commented:
@jli157, do you intend to complete this work?

jli157 commented:
> @jli157, do you intend to complete this work?

I'll try to finish this one. I'm supposing this one will be based on #13364 which is still open. Should I still wait for that one until it's closed? Or just go ahead based on the current code base 2.0?

erwango commented:
@jli157, https://github.com/zephyrproject-rtos/zephyr/pull/13364 is upcoming

erwango commented:
@jli157, fyi new STM32 DMA driver is now merged.

jli157 commented:
> @jli157, fyi new STM32 DMA driver is now merged.

Okay, I'll rework this patch on the latest code base.

FRASTM commented:
@jli157, do you have a new version after rebase ?

FRASTM commented on dts/bindings/serial/st,stm32-uart.yaml:
```diff
@@ -29,4 +29,10 @@ properties:
+    dma-channels:
+      type: array
+      category: optional
+      description: DMA config(slot, tx, rx)  for UART
```
could we have 2 separated arrays: 
>   dma-channel-tx with DMA-config(slot, tx)  
>   dma-channel-rx with DMA-config(slot, rx)

QSQCaito commented:
What would be required to make this go through?

jli157 commented:
> What would be required to make this go through?

Sorry that I was distracted from this one long time and its original base had a lot of changes. I need to read through the current DMA interface changes before working on this one if we still need the changes.

QSQCaito commented:
Thanks for the quick response. Can we help in any way?  This would be quite
useful so I'm glad to contribute.

On Thu, May 14, 2020, 18:57 Jun Li <notifications@github.com> wrote:

> What would be required to make this go through?
>
> Sorry that I was distracted from this one long time and its original base
> had a lot of changes. I need to read through the current DMA interface
> changes before working on this one if we still need the changes.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/zephyrproject-rtos/zephyr/pull/14916#issuecomment-628761877>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGUIF3Z3GP2DLRKSCRRIV43RRQPHLANCNFSM4HBPBMGA>
> .
>

jli157 commented:
> Thanks for the quick response. Can we help in any way? This would be quite useful so I'm glad to contribute.
> […](#)

let me take a look at the base changes to see how to resuscitate it.

erwango commented on soc/arm/st_stm32/stm32f4/dts_fixup.h:
```diff
@@ -111,6 +111,9 @@
 #define DT_UART_STM32_USART_1_CLOCK_BITS	DT_ST_STM32_USART_40011000_CLOCK_BITS
 #define DT_UART_STM32_USART_1_CLOCK_BUS	DT_ST_STM32_USART_40011000_CLOCK_BUS
 #define DT_UART_STM32_USART_1_HW_FLOW_CONTROL	DT_ST_STM32_USART_40011000_HW_FLOW_CONTROL
+#define DT_UART_STM32_USART_1_DMA_SLOT	DT_ST_STM32_USART_40011000_DMA_CHANNELS_0
```
Remove fixups

erwango commented on dts/bindings/serial/st,stm32-uart.yaml:
```diff
@@ -29,4 +29,10 @@ properties:
       description: Set to enable RTS/CTS flow control at boot time
       generation: define
+
+    dma-channels:
```
dma-channels is now part of base.yaml, this could be removed.

erwango commented on dts/bindings/serial/st,stm32-usart.yaml:
```diff
@@ -29,4 +29,10 @@ properties:
       description: Set to enable RTS/CTS flow control at boot time
       generation: define
+
+    dma-channels:
```
dma-channels is now part of base.yaml, this could be removed.

jli157 replied:
Thanks for the hints, @erwango ! I was a bit confused by the following configuration in the dtsi files:
```
                        dmas = <&dma2 3 3 0x400 0x3
                                &dma2 2 3 0x400 0x3>
```

I know the second parameter is the stream no, the third is the channel no. What does the last two parameters mean? Where can I get their descriptions? I have thought `0x400` is the range of the interface registers (got the hint from I2S). However, from the descriptions of the file `vi st,stm32-dma.yaml`, it seems the 4th is `channel-config` and the last one is `features`, right? If so, I guess the configs on I2S could be wrong.

erwango replied:
Parameters description is provided in the dma bindings : `dts/bindings/dma/st,stm32-dma.yaml`

erwango replied:
I let @FRASTM comment on the I2S configuration

jli157 replied:
Never mind. I guess the parameter 0x400 is correct option, which is defined as `increment address between transfers`.

FRASTM replied:
For stm32f4 DMA is V1. First param is the stream (from 0 to 7), second param is the request (aka channels) from 0 to 7 per DMA stream. Example for F411 (according to the ref Man), only DMA1 has requests for I2s (I2S2 or I2S3) 
I2S**2**_EXT_RX  on stream 3 channel request 3     I2S**3**_EXT_RX  on stream 2 channel request 2 
I2S**2**_EXT_TX  on stream 4 channel request 2     I2S**3**_EXT_TX  on stream 5 channel request 2

erwango reviewed (changes requested):
Few hints for rebase.
Mainly I advice to check how is now working dma client in spi_ll_stm32.c

jli157 committed:
dts: uart: enable dma support

Enable UART DMA configurations for all STM32F4 SoCs.

Signed-off-by: Jun Li <jun.r.li@intel.com>

jli157 committed:
uart: stm32: implement async api

Implement asycn UART APIs by enabling DMA transmission
between UART and the data buffer.

* Enabled replacing RX buffer
* If the replacing buffer is same as the current one, no replacement
* Fixed RX buffer counter issue
* Reformatted event function names.
* Make sure the buffer replacement happens only when the current one
  is full.
* Use an inline `async_timer_start` to do timer starting job.
* rebased with c99 types

Signed-off-by: Jun Li <jun.r.li@intel.com>

jli157 commented:
@erwango any idea how I can fetch the DMA driver name for a UART? In the current uart driver initialization macro, only thing I can see is just an "index". I can't figure out how to get DMA properties by only an index number. 

I saw the example on I2S is like this:
```
DT_DMAS_LABEL_BY_NAME(DT_NODELABEL(i2s##index), dir)
```

However, if I do the similar thing for uart:
```
DT_DMAS_LABEL_BY_NAME(DT_NODELABEL(usart##index), dir)
```

It doesn't work since the index numbers passed into the macro are only "0" and "1" for the board "nucleo_f429zi" while the uart devices on the board are actually "usart1" and "usart6".  I don't know how I can get "usart1" and "usart6" by only their index numbers here. 

Also, it seems using "usartx" is not good idea since there are two types of uart ports: "usart" and "uart". 


Any suggestions?

erwango commented:
> @erwango any idea how I can fetch the DMA driver name for a UART? In the current uart driver initialization macro, only thing I can see is just an "index". I can't figure out how to get DMA properties by only an index number.

@jli157, "index" is the index of a device in the list of devices enabled for this driver, starting from 0. 
So, if only one device is enabled (let's say usart2), it will get index "0".
If usart2 and usart6 are enabled, they'll match respectively to index 0 and index 1. To access information you need with DMA DT macros, you should use the DT_INST_DMAS_FOO ones, as for instance:
```
dma_slot = DT_INST_DMAS_CELL_BY_NAME(index, dir, slot)
```
This is the way it is done for SPI DMA client, you can check spi_ll_stm32.c for more details.

jli157 commented:
> > @erwango any idea how I can fetch the DMA driver name for a UART? In the current uart driver initialization macro, only thing I can see is just an "index". I can't figure out how to get DMA properties by only an index number.
> 
> @jli157, "index" is the index of a device in the list of devices enabled for this driver, starting from 0.
> So, if only one device is enabled (let's say usart2), it will get index "0".
> If usart2 and usart6 are enabled, they'll match respectively to index 0 and index 1. To access information you need with DMA DT macros, you should use the DT_INST_DMAS_FOO ones, as for instance:
> 
> ```
> dma_slot = DT_INST_DMAS_CELL_BY_NAME(index, dir, slot)
> ```
> 
> This is the way it is done for SPI DMA client, you can check spi_ll_stm32.c for more details.

Thank you! It works! 

One more question: I'm curious why I can't find any STM32 SPI interfaces are configured with DMA properties?  I searched the directory "soc/st/f4" and couldn't find any SPIs are configured with DMA. Did I search the wrong directory? 


Also I found a problem in `dma_config`. It says that it `cannot configure the dma stream 5`. Thus I checked the file `dma_stm32.c`, found all channels above 4 are conditionally defined with the macros like "DT_INST_IRQ_HAS_IDX(0, 5)". I have no idea where this is defined. it seems to be defined in the DMA dts section. Are they related to `interrupts = <11 0 12 0 13 0 14 0 15 0 16 0 17 0 47 0>;` in the DMA properties?

jli157 commented:
> > > @erwango any idea how I can fetch the DMA driver name for a UART? In the current uart driver initialization macro, only thing I can see is just an "index". I can't figure out how to get DMA properties by only an index number.
> > 
> > 
> > @jli157, "index" is the index of a device in the list of devices enabled for this driver, starting from 0.
> > So, if only one device is enabled (let's say usart2), it will get index "0".
> > If usart2 and usart6 are enabled, they'll match respectively to index 0 and index 1. To access information you need with DMA DT macros, you should use the DT_INST_DMAS_FOO ones, as for instance:
> > ```
> > dma_slot = DT_INST_DMAS_CELL_BY_NAME(index, dir, slot)
> > ```
> > 
> > 
> > This is the way it is done for SPI DMA client, you can check spi_ll_stm32.c for more details.
> 
> Thank you! It works!
> 
> One more question: I'm curious why I can't find any STM32 SPI interfaces are configured with DMA properties? I searched the directory "soc/st/f4" and couldn't find any SPIs are configured with DMA. Did I search the wrong directory?
> 
> Also I found a problem in `dma_config`. It says that it `cannot configure the dma stream 5`. Thus I checked the file `dma_stm32.c`, found all channels above 4 are conditionally defined with the macros like "DT_INST_IRQ_HAS_IDX(0, 5)". I have no idea where this is defined. it seems to be defined in the DMA dts section. Are they related to `interrupts = <11 0 12 0 13 0 14 0 15 0 16 0 17 0 47 0>;` in the DMA properties?

I think I found the reason: the DMA initialization priority is set to "POST_KERNEL" while UART's is "PRE_KERNEL_1".  All drivers using DMA are initialized behind the DMA's. I think the UART driver has to stay in the PRE_KERNEL priority given that it needs to support the board booting up process. I'm wondering if DMA's initialization priority shall be set to be earlier than UART's?

erwango commented:
> One more question: I'm curious why I can't find any STM32 SPI interfaces are configured with DMA properties? I searched the directory "soc/st/f4" and couldn't find any SPIs are configured with DMA. Did I search the wrong directory?

It's configured and used in `tests/drivers/spi/spi_loopback/boards/nucleo_wb55rg.overlay`

> I think I found the reason: the DMA initialization priority is set to "POST_KERNEL" while UART's is "PRE_KERNEL_1". All drivers using DMA are initialized behind the DMA's except UART. I think the UART driver has to stay in the PRE_KERNEL priority given that it needs to support the board booting up process. I'm wondering if DMA's initialization priority shall be set to be earlier than UART's?

It would make sense indeed

QSQCaito commented:
@jli157, thanks for taking this back up!
Please let me know if we can contribute in anyway. We have a custom board with an L4 that will use 3xUARTs with DMA with this async API, so we could help testing for sure. If more implementation is required we can also assist, but it looks like the implementation is already there.
Cheers

jli157 commented:
> @jli157, thanks for taking this back up!
> Please let me know if we can contribute in anyway. We have a custom board with an L4 that will use 3xUARTs with DMA with this async API, so we could help testing for sure. If more implementation is required we can also assist, but it looks like the implementation is already there.
> Cheers

Thanks for the offering! Definitely I need help! I have only several STM32F4 boards. So, you can help me to do the tests on other STM32 boards after I finish the tests on F4 boards to make sure the driver is functioning. 

Right now, I just did some tests for validating DMA parameters, and will do the tests on F4 boards soon.

QSQCaito commented:
Sounds good! I'll wait for your feedback to start testing.

On Tue, May 19, 2020 at 6:52 PM Jun Li <notifications@github.com> wrote:

> @jli157 <https://github.com/jli157>, thanks for taking this back up!
> Please let me know if we can contribute in anyway. We have a custom board
> with an L4 that will use 3xUARTs with DMA with this async API, so we could
> help testing for sure. If more implementation is required we can also
> assist, but it looks like the implementation is already there.
> Cheers
>
> Thanks for the offering! Definitely I need help! I have only several
> STM32F4 boards. So, you can help me to do the tests on other STM32 boards
> after I finish the tests on F4 boards to make sure the driver is
> functioning.
>
> Right now, I just did some tests for validating DMA parameters, and will
> do the tests on F4 boards soon.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/zephyrproject-rtos/zephyr/pull/14916#issuecomment-630947220>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGUIF36FALYVCO2CLRUDLSDRSK2OHANCNFSM4HBPBMGA>
> .
>

jli157 commented:
> Sounds good! I'll wait for your feedback to start testing.
> […](#)

Meanwhile, can you help to modify the corresponding L4's dtsi files to add the DMA support for those UART ports?

QSQCaito commented:
Yes, we can do that.

On Tue, May 19, 2020 at 6:57 PM Jun Li <notifications@github.com> wrote:

> Sounds good! I'll wait for your feedback to start testing.
> … <#m_-3116132168399613793_>
>
> Meanwhile, can you help to modify the corresponding L4 SoC's dtsi files to
> add the DMA support for those UART ports?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/zephyrproject-rtos/zephyr/pull/14916#issuecomment-630949665>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGUIF3ZVQSEOZNJ5UYMRBITRSK27XANCNFSM4HBPBMGA>
> .
>

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -751,6 +1290,16 @@ static void uart_stm32_irq_config_func_##index(struct device *dev)	\
 
+#ifdef CONFIG_UART_ASYNC_API
+#define STM32_UART_DMA_DATA(index)					\
+	UTIL_AND(DT_INST_DMAS_HAS_NAME(index, rx),			\
```
Can you check https://github.com/zephyrproject-rtos/zephyr/pull/25286 ?
It will allow more flexible use of channels

jli157 replied:
Thanks for the comments! I'll refer to that changes and address your comments.

jli157 replied:
Actually, I'm a bit confused. My current implementation does support disabling DMA since the STM32_UART_DMA_DATA() will be nothing if `CONFIG_UART_ASYNC_API` is not defined. What do you mean allowing more flexible use of channels? I've checked #25286 and it seems to do the similar implementation to mine.

erwango replied:
Aim of #25286 is to allow use and definition (in dts nodes) of only one DMA channel Tx or Rx.
With current implementation, there is one case where this won't compile.

jli157 replied:
I see, will check it later.

erwango reviewed (changes requested):
Nice (re)start!
A couple of quick indications

jli157 committed:
samples: uart_async: implement uart async test

For testing purpose, it could be removed after the PR
is finished.

Signed-off-by: Jun Li <jun.r.li@intel.com>

QSQCaito commented:
> > Sounds good! I'll wait for your feedback to start testing.
> > […](#)
> 
> Meanwhile, can you help to modify the corresponding L4 SoC's dtsi files to add the DMA support for those UART ports?

@jli157 two questions:
1. Should we push the .dtsi files to this PR?
2. _(repeat question from Slack, but maybe you know)_  How can we handle a USART that can take multiple DMA streams?  For example for USART1 we can either use DMA1 channels 2/3 or DMA2 channels 6/7.  How would we represent this in the .dtsi ? We put just one?

QSQCaito commented:
> > > Sounds good! I'll wait for your feedback to start testing.
> > > […](#)
> > 
> > 
> > Meanwhile, can you help to modify the corresponding L4 SoC's dtsi files to add the DMA support for those UART ports?
> 
> @jli157 two questions:
> 
> 1. Should we push the .dtsi files to this PR?
> 2. _(repeat question from Slack, but maybe you know)_  How can we handle a USART that can take multiple DMA streams?  For example for USART1 we can either use DMA1 channels 2/3 or DMA2 channels 6/7.  How would we represent this in the .dtsi ? We put just one?

Question 2. got answered in Slack: 
> put one config and the other one under comment ? Anyway, application will still have option to >overwrite the property and chose the alternate config, so this would just be for documentation.

jli157 commented:
> > > Sounds good! I'll wait for your feedback to start testing.
> > > […](#)
> > 
> > 
> > Meanwhile, can you help to modify the corresponding L4 SoC's dtsi files to add the DMA support for those UART ports?
> 
> @jli157 two questions:
> 
> 1. Should we push the .dtsi files to this PR?
> 2. _(repeat question from Slack, but maybe you know)_  How can we handle a USART that can take multiple DMA streams?  For example for USART1 we can either use DMA1 channels 2/3 or DMA2 channels 6/7.  How would we represent this in the .dtsi ? We put just one?

For your Question 1, maybe you can pull a new branch from mine and keep working on your own? later we can merge together to this one so that you will be the co-authors? I think working together on the same branch could be difficult for us since we could create conflicts time to time. Another option is that you can wait for this one is finished for F4 series and then push your new changes behind my commits or just create a new PR as an improvement for the whole STM32 SoCs. I guess most likely I would be requested to increase the coverage of this PR to all STM32 SoCs. :)

jli157 commented:
The commit c489ceb comes form #25492, and will be removed once the PR is approved and merged.

jli157 committed:
board: nucleo_f429zi: enable dma

enable two dmas to support async uart.

Signed-off-by: Jun Li <jun.r.li@intel.com>

erwango commented on dts/bindings/serial/st,stm32-uart.yaml:
```diff
@@ -16,3 +16,9 @@ properties:
       description: Configures the parity of the adapter. Value 0 for none, 1 for odd and 2 for even parity.
       default: 0
+
+    dmas:
+      required: false
+
+    dma-names:
+      required: false
```
You can just remove these as they are already included in base.yaml binding.
(sorry if this was not clear enough in the initial comment)

erwango commented on dts/bindings/serial/st,stm32-usart.yaml:
```diff
@@ -16,3 +16,9 @@ properties:
       description: Configures the parity of the adapter. Value 0 for none, 1 for odd and 2 for even parity.
       default: 0
+
+    dmas:
```
You can just remove these as they are already included in base.yaml binding.
(sorry if this was not clear enough in the initial comment)

jli157 replied:
Sure.

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -643,6 +1119,14 @@ static const struct uart_driver_api uart_stm32_driver_api = {
 	.irq_callback_set = uart_stm32_irq_callback_set,
 #endif	/* CONFIG_UART_INTERRUPT_DRIVEN */
+#ifdef CONFIG_UART_ASYNC_API
+	.callback_set = uart_stm32_callback_set,
+	.tx = uart_stm32_tx,
+	.tx_abort = uart_stm32_tx_abort,
+	.rx_enable = uart_stm32_rx_enable,
+	.rx_disable = uart_stm32_rx_disable,
+	.rx_buf_rsp = uart_stm32_rx_buf_rsp,
```
We might want to make clear these functions are async/dma dedicated.
So maybe `uart_stm32_tx` should be renamed to `uart_stm32_async_tx` and so on ....

jli157 replied:
will change to reflect that.

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -724,14 +1208,58 @@ static int uart_stm32_init(struct device *dev)
+
+	int ret = 0;
+#ifdef CONFIG_UART_ASYNC_API
+	ret = uart_stm32_async_init(dev);
```
Please make sure that it will be possible to have mixed use of async and non async uart devices.

jli157 replied:
I think once both CONFIG_UART_ASYNC_API and CONFIG_UART_INTERRUPT_DRIVEN are enabled, every UART port will have both capabilities. 

I tried the build without CONFIG_UART_ASYNC_API enabled, no complication errors. anyway, I'll check the conditions time to time to make sure it doesn't break.

jli157 commented:
@erwango by the way, to make review easier, I think it could be better I submit new commits for further changes for a specific file; and once everything is done, I can squash them to be several independent commits?

erwango commented:
> @erwango by the way, to make review easier, I think it could be better I submit new commits for further changes for a specific file; and once everything is done, I can squash them to be several independent commits? or I still keep amending them?

Since this is still a WIP, I'm fine with this proposal, if this is easier to you

jli157 commented:
Pushed new changes, addressing the comments to yaml files and the commit "uart: stm32: implement async api". Here I'm committing a new commit as an amending one which will be squashed later.

jli157 commented:
@erwango Do you have idea how to fix this format issue? 
```
-:116: ERROR:SPACING: need consistent spacing around '*' (ctx:WxV)
#116: FILE: drivers/serial/uart_stm32.c:631:
+	USART_TypeDef *UartInstance = UART_STRUCT(dev);
```

That line was copied from another place. I'm not sure what is wrong with it.

erwango commented:
> USART_TypeDef

This is not obvious to me...

jli157 commented:
## TX and RX are verified working with the echo test.

#### Host side:
```
 $ python3 test_uart.py /dev/ttyUSB0 

Write: This is a stm32 F429zi packet!
rcv: b'This is a stm32 F429zi packet!\r\n'

Write: This is a stm32 F429zi packet!
rcv: b'This is a stm32 F429zi packet!\r\n'

Write: This is a stm32 F429zi packet!
rcv: b'This is a stm32 F429zi packet!\r\n'

Write: This is a stm32 F429zi packet!
rcv: b'This is a stm32 F429zi packet!\r\n'
```

#### Device Side
```
RX done: 32 - This is a stm32 F429zi packet!

TX done: 0
RX done: 32 - This is a stm32 F429zi packet!

TX done: 0
RX done: 32 - This is a stm32 F429zi packet!

TX done: 0
RX done: 32 - This is a stm32 F429zi packet!

TX done: 0
RX done: 32 - This is a stm32 F429zi packet!

TX done: 0

```

jli157 committed:
test: uart_async_api: add board nucleo_f429zi

Define UART_DEV_NAME as "UART_6" if the board
nucleo_f429zi is selected.

Signed-off-by: Jun Li <jun.r.li@intel.com>

jli157 commented:
* Removed the temporary uart_asyc test. 
* Added nucleo_f429zi on the test board lists for `tests/drivers/uart/uart_async_api`
* Ran the test cases, most cases still failed. :(

jli157 commented:
@erwango and @QSQCaito It seems the design idea of test cases in `tests/drivers/uart/uart_async_api` doesn't suppose that RX is using the circular buffer to reuse the same RX buffer. If the chained RX is required, the message `UART_RX_BUF_REQUEST` will be sent to ask the client to set the next RX buffer. Here, my design is trying to reuse the same RX buffer to continuously receive data. Looks like I have to give up using the circular buffer mode for implementing RX reception.

jli157 committed:
test: uart_async_api: fix chained read bug

chained read buffer needs to rotate.

Signed-off-by: Jun Li <jun.r.li@intel.com>

erwango commented:
> @erwango and @QSQCaito It seems the design idea of test cases in `tests/drivers/uart/uart_async_api` doesn't suppose that RX is using the circular buffer to reuse the same RX buffer. If the chained RX is required, the message `UART_RX_BUF_REQUEST` will be sent to ask the client to set the next RX buffer. Here, my design is trying to reuse the same RX buffer to continuously receive data so as to save the time on reloading DMA settings, which could be critical for timing sensitive UART RX reception. Looks like I have to give up using the circular buffer mode for implementing RX reception to pass all test cases.

@Mierunski, as codeowner, can you comment on implementation option proposed by @jli157 ?

Mierunski commented:
API requires that you can provide second buffer to receive data after first is filled and that is checked in use cases, it has to be possible to change buffer, not only use single circular buffer.
Using circular buffer gives possibility that data passed to user on UART_RX_RDY event can be altered by driver while it is processed.
uart_rx_buf_rsp has a note saying that providing buffer that is already in usage can lead to undefined behaviour, i.e. data can be overwritten. But if someone intends to use single circular buffer this could be the option. You can check if buffer provided as response to UART_RX_BUF_REQUEST is the same as already set, if it is then enable circular DMA, if not then change buffer when first one is full.

jli157 commented:
> API requires that you can provide second buffer to receive data after first is filled and that is checked in use cases, it has to be possible to change buffer, not only use single circular buffer.
> Using circular buffer gives possibility that data passed to user on UART_RX_RDY event can be altered by driver while it is processed.
> uart_rx_buf_rsp has a note saying that providing buffer that is already in usage can lead to undefined behaviour, i.e. data can be overwritten. But if someone intends to use single circular buffer this could be the option. You can check if buffer provided as response to UART_RX_BUF_REQUEST is the same as already set, if it is then enable circular DMA, if not then change buffer when first one is full.
> Of course this is still labeled as undefined behaviour as it is vendor dependent.

Thanks for the insight sharing! Very helpful! Right now my implementation is like what you said: if the next buffer is not set and the current one is full, it will behave as a circular one; if the next one is set, it will replace the current one with the next one. 

The `uart_async_test` is well designed. I'll make sure it can go through all test cases.

jli157 commented:
@QSQCaito I think the implementation is almost done except some bugfix work is left. You can proceed to do the tests on your platform. 

I found very strange bug when running `uart_async_test` as follows:
```
starting test - test_chained_read
replace buffer (20)


ch-R: 0
tx: l=10
DMA
tx done:0
idle
rx_rdy: (0 10)
Replacing RX buffer: 20
replace buffer (30)
read_len: 10, tx_buf_sz: 10


ch-R: 1
tx: l=10
DMA
tx done:0
idle
rx_rdy: (0 10)
read_len: 10, tx_buf_sz: 10


ch-R: 2
tx: l=10
DMA
tx done:0
idle
rx_rdy: (10 20)
Replacing RX buffer: 30
replace buffer (10)
read_len: 10, tx_buf_sz: 10


ch-R: 3
tx: l=10
DMA
tx done:0
idle
rx_rdy: (0 20)
read_len: 20, tx_buf_sz: 10


ch-R: 4
tx: l=10
DMA
tx done:0
idle
rx_rdy: (0 10)
read_len: 10, tx_buf_sz: 10

    Assertion failed at ../src/test_uart_async.c:169: test_chained_read: (memcmp(tx_buf, read_ptr, sizeof(tx_buf)) not equal to 0)
Buffers not equal
FAIL - test_chained_read

```

In the chained reading test, there are six reading operations to read 10 bytes every time. However, on the round 3 right after the last buffer was replaced, uart RX read 20 bytes instead of 10 bytes, which is really puzzling me. I guess maybe the DMA RX is not well reset after the buffer replacement happens. But I can't find the root cause. Maybe you can do the same test on your platform to see if the behaviors are same? 

Thank you!

QSQCaito commented:
> @QSQCaito I think the implementation is almost done except some bugfix work is left. You can proceed to do the tests on your platform.
> 
> I found very strange bug when running `uart_async_test` as follows:
> 
> ```
> starting test - test_chained_read
> replace buffer (20)
> 
> 
> ch-R: 0
> tx: l=10
> DMA
> tx done:0
> idle
> rx_rdy: (0 10)
> Replacing RX buffer: 20
> replace buffer (30)
> read_len: 10, tx_buf_sz: 10
> 
> 
> ch-R: 1
> tx: l=10
> DMA
> tx done:0
> idle
> rx_rdy: (0 10)
> read_len: 10, tx_buf_sz: 10
> 
> 
> ch-R: 2
> tx: l=10
> DMA
> tx done:0
> idle
> rx_rdy: (10 20)
> Replacing RX buffer: 30
> replace buffer (10)
> read_len: 10, tx_buf_sz: 10
> 
> 
> ch-R: 3
> tx: l=10
> DMA
> tx done:0
> idle
> rx_rdy: (0 20)
> read_len: 20, tx_buf_sz: 10
> 
> 
> ch-R: 4
> tx: l=10
> DMA
> tx done:0
> idle
> rx_rdy: (0 10)
> read_len: 10, tx_buf_sz: 10
> 
>     Assertion failed at ../src/test_uart_async.c:169: test_chained_read: (memcmp(tx_buf, read_ptr, sizeof(tx_buf)) not equal to 0)
> Buffers not equal
> FAIL - test_chained_read
> ```
> 
> In the chained reading test, there are six reading operations to read 10 bytes every time. However, on the round 3 right after the last buffer was replaced, uart RX mistakenly reported to have read 20 bytes instead of the normal 10 bytes, which is really puzzling me. I guess maybe the DMA RX is not well reset after the buffer replacement happens. But I can't find the root cause. Maybe you can do the same test on your platform to see if the behaviors are same?
> 
> By the way, the test is a UART self loop test. You need to connect the TX pin of the UART port you are choosing to its RX pin.
> 
> Thank you!

Awesome, thanks!
I will proceed with the tests late this week or early next and report back as soon as I can. Tomorrow we will create a PR for the .dtsi of our boards.

jli157 committed:
uart_stm32: debug tools

Signed-off-by: Jun Li <jun.r.li@intel.com>

giansta commented:
Hello @jli157 
I'm trying to use this PR code to run it on our stm32l462 platform. We have four serial lines for which we defined the dma resource in dtsi files in a similar way as you did. Just a minor change was needed in dma_stm32.c, as the L4 family has two distinct TX and RX data registers.
I just started testing, yet something doesn't work. I'm running the tests/drivers/uart/uart_async_api and this is showing logs forever from the drivers and never hitting the comparison of TX and RX buffer in first test test_single_read(). Most probably I'm doing some mistake, I need some more basic test to run and understand better the code. Yet maybe you could have the right hint. Thanks.

giansta commented:
Hello @jli157 
I'm trying to use this PR code to run it on our stm32l462 platform. We have four serial lines for which we defined the dma resource in dtsi files in a similar way as you did. Just a minor change was needed in dma_stm32.c, as the L4 family has two distinct TX and RX data registers.
I just started testing, yet something didn't work. I'm running the tests/drivers/uart/uart_async_api and this is showing logs forever from the drivers and never hitting the comparison of TX and RX buffer in first test test_single_read(). Moreover, the logs show only RX events forever instead of the TX events I was expecting first.

`Running test suite uart_async_test
===================================================================
START - test_single_read_setup
 PASS - test_single_read_setup
===================================================================
START - test_single_read
DMA 5/7 0 0 1 1             id=5 max_streams=7 source_periph=0 (mem) busy=0 src_size=dst_size=1
rx_rdy: (0 1) 00
DMA 5/7 0 0 1 1
rx_rdy: (1 5) 00 00 00 00 00
DMA 5/7 0 0 1 1
rx_rdy: (5 5) 00 00 00 00 00
DMA 5/7 0 0 1 1
rx_rdy: (0 4) 00 00 00 00
DMA 5/7 0 0 1 1
rx_rdy: (0 2) 00 00
DMA 5/7 0 0 1 1
rx_rdy: (2 8) 00 00 00 00 00 00 00 00
DMA 5/7 0 0 1 1
rx_rdy: (0 3) 00 00 00`

(Here I tried to add some info in the prints as added in the 1st line, icluding the RX data that don't seem real).
I'm using a serial line that is connected to another microcontroller which implements the loopback, yet this is not receiving at moment.
Most probably I'm doing some mistake, I need some more basic test to run and understand better the code. Yet maybe you 
could have the right hint. Thanks.

jli157 commented:
> Hello @jli157
> I'm trying to use this PR code to run it on our stm32l462 platform. We have four serial lines for which we defined the dma resource in dtsi files in a similar way as you did. Just a minor change was needed in dma_stm32.c, as the L4 family has two distinct TX and RX data registers.
> I just started testing, yet something didn't work. I'm running the tests/drivers/uart/uart_async_api and this is showing logs forever from the drivers and never hitting the comparison of TX and RX buffer in first test test_single_read(). Moreover, the logs show only RX events forever instead of the TX events I was expecting first.
> 
> `START - test_single_read DMA 5/7 0 0 1 1 rx_rdy: (0 1) 00 DMA 5/7 0 0 1 1 rx_rdy: (1 5) 00 00 00 00 00 DMA 5/7 0 0 1 1 rx_rdy: (5 5) 00 00 00 00 00 DMA 5/7 0 0 1 1 rx_rdy: (0 4) 00 00 00 00 DMA 5/7 0 0 1 1 rx_rdy: (0 2) 00 00 DMA 5/7 0 0 1 1 rx_rdy: (2 8) 00 00 00 00 00 00 00 00 DMA 5/7 0 0 1 1 rx_rdy: (0 3) 00 00 00`
> 
> Here I tried to add some info in the prints. Lines with "DMA" report id=5 max_streams=7 source_periph=0 (mem) busy=0 src_size=dst_size=1; the RX lines print data that don't seem real.
> I'm using a serial line that is connected to another microcontroller which implements the loopback, yet this is not receiving at moment.
> Most probably I'm doing some mistake, I need some more basic test to run and understand better the code. Yet maybe you
> could have the right hint. Thanks.

Hi The loopback test is conducted actually on the same port of the same board. You need to connect the TX pin to the RX pin of the same UART port before running the tests. 

Anyway, this PR has not fully gone through the uart async test cases. Some cases fails without obvious reasons. 

Another test it could be helpful to you is the newest patch I just uploaded (7e1609f87f6e7ca217c0be84a25b3ff8e8bd3a21): you can use a FTDI USB-UART cable to connect your computer and the board, and run the python script inside the test directory to see if it works.

giansta commented:
Hi @jli157 
Thank you for your additional test, I believe it will be very helpful.
In the meantime I was testing both with your uart_async_api test and with our application, where I could start the TX either in interrupt or DMA mode.
Using  two apps I could find some mistakes I did in porting to STM32L4 and doing a small progress. I'm reporting now few observations.
1. Using our app I have now debugging issues, both using eclipse and gdb. This didn't happen before using the zephyr branch with async uart. Yet debugging works when using uart_async_api test, I have to find out why.
2. Using our app I can see the waveform of the 5-byte transmission, yet it's bit different in the last part and a bit shorter than when transmitting with interrupt.
3. I'm keeping using the other processor as receiver, yet we're using HW flow control in the BSP. Using our app and I see that the RTS is always high (inactive) when using DMA, preventing the other processor to reply; instead it's well handled when using interrupt and the other processor replies well with loopback echo.
4. Using the uart_async_api test with HW flow control, the RTS is always low (active), but nothing appears on the TX pin, even if the TX_DONE event comes.
I'm investigating on these points and in parallel I'm going to remove the HW flow control now to test a simplified setup.
Any your suggestion is welcome. Thanks!

giansta commented:
> @QSQCaito I think the implementation is almost done except some bugfix work is left. You can proceed to do the tests on your platform.
> 
> I found very strange bug when running `uart_async_test` as follows:
> 
> In the chained reading test, there are six reading operations to read 10 bytes every time. However, on the round 3 right after the last buffer was replaced, uart RX mistakenly reported to have read 20 bytes instead of the normal 10 bytes, which is really puzzling me. I guess maybe the DMA RX is not well reset after the buffer replacement happens. But I can't find the root cause. Maybe you can do the same test on your platform to see if the behaviors are same?
> 
> By the way, the test is a UART self loop test. You need to connect the TX pin of the UART port you are choosing to its RX pin.
> 
> Thank you!

Hi @jli157 
I believe to have found the reason for the chained reading test fails. It's because of usage of "number of data to transfer register" (DMA_CNDTRx) as the number of received bytes: indeed the dma_stm32_get_status() (and LL_DMA_GetDataLength()) are returning the number of bytes "to be transmitted".
So, in the round 3, the 20 bytes are the remaining bytes, while received bytes are 30-20=10.
In the previous rounds the things work the same, either because the circular buffer mode reloads the value zero with the buffer length or because the received and remaining bytes are same (e.g. 5=10-5).
At moment I'm trying to fix this without circular buffer mode (testing the L4 family platform that @QSQCaito was mentioning). I still need to check the other test cases. 
Then we should figure out if there is a way to use the keep using the circular buffer mode.
I'll put my changes in another branch as son as they work better. In the meantime could you please check if this is right or not? Thanks.

jli157 commented:
> > @QSQCaito I think the implementation is almost done except some bugfix work is left. You can proceed to do the tests on your platform.
> > I found very strange bug when running `uart_async_test` as follows:
> > In the chained reading test, there are six reading operations to read 10 bytes every time. However, on the round 3 right after the last buffer was replaced, uart RX mistakenly reported to have read 20 bytes instead of the normal 10 bytes, which is really puzzling me. I guess maybe the DMA RX is not well reset after the buffer replacement happens. But I can't find the root cause. Maybe you can do the same test on your platform to see if the behaviors are same?
> > By the way, the test is a UART self loop test. You need to connect the TX pin of the UART port you are choosing to its RX pin.
> > Thank you!
> 
> Hi @jli157
> I believe to have found the reason for the chained reading test fails. It's because of usage of "number of data to transfer register" (DMA_CNDTRx) as the number of received bytes: indeed the dma_stm32_get_status() (and LL_DMA_GetDataLength()) are returning the number of bytes "to be transmitted".
> So, in the round 3, the 20 bytes are the remaining bytes, while received bytes are 30-20=10.
> In the previous rounds the things work the same, either because the circular buffer mode reloads the value zero with the buffer length or because the received and remaining bytes are same (e.g. 5=10-5).
> At moment I'm trying to fix this without circular buffer mode (testing the L4 family platform that @QSQCaito was mentioning). I still need to check the other test cases.
> Then we should figure out if there is a way to keep using the circular buffer mode.
> I'll put my changes in another branch as son as they work better. In the meantime could you please check if this diagnosis is right or not? Thanks.

@giansta Thanks for the findings! Yes, please upload your branch so that I can take a look. And sorry for late reply. Got busy on other stuffs recently.

giansta commented:
> @giansta Thanks for the findings! Yes, please upload your branch so that I can take a look. And sorry for late reply. Got busy on other stuffs recently.

Hi @jli157
I have pushed the main changes in my branch https://github.com/giansta/zephyr/commits/dma_uart_stm32_gs
The fix I was talking about is in commit b4c30f7. It would be nice if you could check if this works the same also for F4 family.
I'm cleaning my code and I'll push few other commits.
Currently the test_chained_read is passing, but the test_double_buffer and following ones are still failing.
Anyway I'm checking another fix that I believe should work at least for test_double_buffer case.

giansta commented:
@jli157
I've pushed few other fixes in my branch.
The latest one, 64fd666, fixes the double buffer test. Unfortunately the following ones are still failing, hopefully we'll fix them soon.
The commit 94aa7e3 is intented to fix a minor bug and simplify the code. I found it while debugging, as the following line looked wrong to me.
```
stream->source_periph = stream->direction == MEMORY_TO_PERIPHERAL;
```
I changed it into PERIPHERAL_TO_MEMORY and my test were running the same, most probably because source_burst_length or dest_burst_length are normally the same in my use cases. Anyway, it would be nice if you could confirm that it's not breaking the test in your platform.

jli157 commented:
> @jli157
> I've pushed few other fixes in my branch.
> The latest one, [64fd666](https://github.com/zephyrproject-rtos/zephyr/commit/64fd666489dd65023c582e911ddc08a594f5a47f), fixes the double buffer test. Unfortunately the following ones are still failing, hopefully we'll fix them soon.
> The commit [94aa7e3](https://github.com/zephyrproject-rtos/zephyr/commit/94aa7e3112726ce2c7aef73c3ea328e5fa3b1aad) is intented to fix a minor bug and simplify the code. I found it while debugging, as the following line looked wrong to me.
> 
> ```
> stream->source_periph = stream->direction == MEMORY_TO_PERIPHERAL;
> ```
> 
> I changed it into PERIPHERAL_TO_MEMORY and my test were running the same, most probably because source_burst_length or dest_burst_length are normally the same in my use cases. Anyway, it would be nice if you could confirm that it's not breaking the test in your platform.

Thank you! I'll take a look at your patches? Can I merge parts of them (or maybe full) into this PR if they do fix the same issues on F4 as well? I'd like to ask you to be the co-authors of this PR.

giansta commented:
> Thank you! I'll take a look at your patches? Can I merge parts of them (or maybe full) into this PR if they do fix the same issues on F4 as well? I'd like to ask you to be the co-authors of this PR.

It would be perfect, please merge and/or amend them if needed. Thanks!

giansta commented:
@jli157 I have rebased my branch and added some commit.

Now six tests of eight are passing on our platform.
About the two failing test, I don't understand why the test_long_buffers expects to send 1000 bytes and have the first RX ready event after 524 bytes and the other at the end. This seems a behaviour of some given platform, not the general case, in my opinion.
I still have to investigate why the last one, test_forever_timeout, is failing, I gave up at moment.

I rather tried to use again the Circular buffer mode. 
My conclusion is that it's not compatible with the double buffer handling, because, when the buffer is filled, the DMA uses the buffer circularly without any ISR that can allow a buffer replace. So we should rather use a single buffer and don't generate any UART_RX_BUF_REQUEST and UART_RX_BUF_RELEASED events.
In my opinion the solution should imply a new config option or a change in the API.

Another point is the errors handling. Now our board is using the HW flow control and everything seems to work fine.
Yet, without flow control and using the Chained buffer mode, when the buffer is filled and the replace buffer process is happening (as you were mentioning, using Circular buffer mode would be more efficient and avoid this), an overrun error happens.
I patched this clearing the UART flag, but most probably we need to pass whatever error (overrun, parity, frame) to the uart_stm32_dma_rx_cb() to have a more robust driver.
This change shouldn't be difficult, I can try it out later on.

wayToGoPotato commented:
@jli157 I was wondering if you are still working on this thread?
